server/src/main/java/org/apache/druid/segment/realtime/appenderator/SinkQuerySegmentWalker.java

```diff
@@ -196,13 +200,26 @@ public <T> QueryRunner<T> getQueryRunnerForSegments(final Query<T> query, final
     final LinkedHashMap<SegmentDescriptor, List<QueryRunner<T>>> allRunners = new LinkedHashMap<>();
 
     try {
-      for (final SegmentDescriptor newDescriptor : specs) {
-        final SegmentDescriptor descriptor = newIdToBasePendingSegment.getOrDefault(newDescriptor, newDescriptor);
-        final PartitionChunk<Sink> chunk = sinkTimeline.findChunk(
+      for (final SegmentDescriptor descriptor : specs) {
+        final PartitionChunk<SegmentDescriptorOvershadowable> descriptorChunk = upgradeDescriptorTimeline.findChunk(
```

Comment:
What sort of issues would arise without using this timeline?

Reply:
The timeline is needed to get all the segments corresponding to a given version. There can be missing segments if this is not used.

At this particular point in code, this is quite equivalent to the existing code where we find a sink using `sinkTimeline.findChunk(segmentDescriptor)` to find an object that encompasses a given interval for a fixed <version, partitionNumber>